### ISSUE TITLE
How to Record Crypto Fee Income with Cost Basis

### ISSUE THREAD

**Issue by ArthoPacini** [beancount/beancount#888]
Hi there,

I’m having trouble representing a trading fee scenario in Beancount, **from the exchange perspective**, where part of the fee is collected in cryptocurrency. Here’s the simplified scenario:

1. Alice and Bob match on an order:  
   - Alice sells 1 BTC at 100,000.00 USD.  
   - Bob buys 1 BTC at 100,000.00 USD.

2. Both pay a 1% fee. Instead of Alice receiving 100,000.00 USD, she gets 99,000.00 USD, and the exchange takes 1,000.00 USD as a fee. Similarly, Bob receives 0.99 BTC instead of 1 BTC, and the exchange takes 0.01 BTC as a fee.

The problem:  
- The exchange ends up with a fee of 1000.00 USD and 0.01 BTC.  
- I need to record this 0.01 BTC as income (part of the fee the exchange earned) and also assign it a cost basis.  
- The cost basis should reflect the trade price at the time (100,000.00 USD per BTC), so 0.01 BTC is effectively acquired by the exchange at a cost basis of 1,000.00 USD.  
- I want to ensure this 0.01 BTC is shown in the exchange’s inventory with a cost annotation so that future gains/losses on this BTC can be calculated correctly.  
- However, I’m having trouble getting Beancount entries right so that the 0.01 BTC is recognized as income (in-kind) and simultaneously attaches a 1,000.00 USD cost basis without causing any balancing issues.

Below is a snippet of the Beancount entries I have so far. The trade execution lines show the fees as negative amounts in Income:Exchange:Fees, but I need help figuring out how to properly restructure the entries so that the exchange ends up with something like:

```beancount

option "title" "Crypto Exchange"
option "name_assets" "Assets"
option "name_liabilities" "Liabilities"
option "name_equity" "Equity"
option "name_income" "Income"
option "name_expenses" "Expenses"
option "operating_currency" "USD, BTC"
option "render_commas" "TRUE"
plugin "beancount.plugins.auto_accounts"
option "operating_currency" "USD"

2024-10-03 * "Alice deposits 5.0 BTC (Bitcoin)"
  #deposit
    Assets:Crypto:Custody                        5.00000000 BTC
    Liabilities:Users:Alice:Wallets:Available:0  -5.00000000 BTC
    ; Transaction Type: Deposit
    ; User ID: Alice
    ; Sub-Wallet ID: 0
    ; From: 3Q9St1somesXHAs7eZ9ScE7jYWhdMtkXL
    ; To Address: bc1qxy2addreygjrsqtzq2n0yrf2493p83kkfjhx0wlh
    ; Tx ID: 0441f57d93f82553432542536808f1552d247aa3b1d59d7d6377015aa33484440
    ; Timestamp: 2024-10-03 10:00:00 UTC

2024-10-03 * "Bob deposits 100,000.00 USD via bank transfer"
  #deposit
    Assets:Fiat:Custody                           100000.00 USD
    Liabilities:Users:Bob:Wallets:Available:0    -100000.00 USD
    ; Transaction Type: Deposit
    ; User ID: Bob
    ; Sub-Wallet ID: 0
    ; From: Banking ABC
    ; Transfer ID: PIX123456789
    ; Memo: N/A
    ; Timestamp: 2024-10-03 10:10:00 UTC

2024-10-03 * "Alice places limit sell order #1: 1 BTC @ 100,000 USD"
  #create_order
    Liabilities:Users:Alice:Wallets:Available:0  1.00000000 BTC
    Liabilities:Users:Alice:Wallets:Frozen      -1.00000000 BTC
    ; Transaction Type: Order Placement
    ; Order ID: 1
    ; User ID: Alice
    ; Sub-Wallet ID: 0
    ; Market: BTC/USD
    ; Type: LIMIT SELL
    ; Price: 100,000.00 USD
    ; Quantity: 1.00000000 BTC
    ; Timestamp: 2024-10-03 10:15:00 UTC

2024-10-03 * "Bob places limit buy order #2: 1 BTC @ 100,000 USD"
  #create_order
    Liabilities:Users:Bob:Wallets:Available:0     100000.00 USD
    Liabilities:Users:Bob:Wallets:Frozen         -100000.00 USD
    ; Transaction Type: Order Placement
    ; Order ID: 2
    ; User ID: Bob
    ; Sub-Wallet ID: 0
    ; Market: BTC/USD
    ; Type: LIMIT BUY
    ; Price: 100,000.00 USD
    ; Quantity: 1.00000000 BTC
    ; Timestamp: 2024-10-03 10:20:00 UTC

2024-10-03 * "Trade executed between order #1 (Alice) and order #2 (Bob)"
  #trade
    Liabilities:Users:Alice:Wallets:Frozen       1.00000000 BTC
    Liabilities:Users:Bob:Wallets:Frozen          100000.00 USD
    Liabilities:Users:Alice:Wallets:Available:0   -99900.00 USD
    Liabilities:Users:Bob:Wallets:Available:0   -0.99900000 BTC
    ; Record fees as income
    Income:Exchange:Fees                        -0.00100000 BTC
    Income:Exchange:Fees                            -100.00 USD
    ; Transaction Type: Trade Execution
    ; Buy Order ID: 2
    ; Sell Order ID: 1
    ; Price: 100,000.00 USD
    ; Quantity: 1.00000000 BTC
    ; Fees: 0.1% charged to both parties
    ; Timestamp: 2024-10-03 10:25:00 UTC

```

On the final transaction, I can't use:

```beancount
Income:Exchange:Fees                        -0.00100000 BTC {100000.00 USD} ; Does not balance
```

I want the final accounting to reflect that the exchange earned both USD and BTC as fees, with the BTC recorded at the market price at the time.

How can I properly record this scenario in Beancount so that the 0.01 BTC fee income is booked into the exchange’s inventory with the correct cost basis, and the total fee income (in USD and BTC) is correctly reflected in the Income accounts?

Would a conversion transaction/accounts work here?

**Comment by ArthoPacini** [comment]
Alright, got it...

It's simple actually.

When an user performs an deposit, the exchange assets is increased, but also the liability, because the exchange owes the user.

Alice deposits 5.0 BTC, the exchange assets under custody increase by 5.0 BTC, but also it's liabilities into -5.0 BTC,
Same goes for Bob, he deposits 200,000.00 USD, now under custody as assets the exchange holds, but the exchange also owes him 200k USD.

Then, the funds are locked because both place orders. It just change accounts to better reflect where the funds goes, just signaling they are now locked.

When the trade occurs, if we take the fee away, what actually happens is that the users 'swap' the liability the exchange owes them (or exchange)... so the Liability of Alice goes to Bob and the Liability of Bob goes to Alice.

Now the exchange owes Alice 4.0 Bitcoins and 100k USD, as well as 1 BTC and 100k USD for bob. Note that the actual assets held by the exchange continues the same.

If we add the fee in, then we swap the liability around, but decrease it.

Now Alice owns 99k USD and bob owns 0.99 BTC.

1k USD from Alice fee goes into Income, as well as 1k USD from bob, but now the assets move from custody of the users to the exchange custody.

```beancount

2024-10-03 * "Trade executed between order #1 (Alice) and #2 (Bob)"
  #trade
  ; Now the orders match and a trade is executed:
  ; Alice sells 1 BTC for 100,000 USD, Bob buys 1 BTC for 100,000 USD.
  ; However, the exchange charges fees. Assume a 1% total fee split: 
  ; Alice receives 99,000 USD instead of 100,000 USD (so 1,000 USD fee taken from her side),
  ; Bob receives 0.99 BTC instead of 1 BTC (0.01 BTC fee taken from his side).
  ; The total fees collected by the exchange are: 1,000 USD + 0.01 BTC (worth 1,000 USD at the trade price).
  ; Below we record all these changes:
  ; First, clear the frozen funds since the order is executed:
  Liabilities:Users:Alice:Wallets:Frozen         1.00000000 BTC
  ; Alice’s frozen BTC is released from the order and will now be reallocated.
  Liabilities:Users:Bob:Wallets:Frozen            100000.00 USD
  ; Bob’s frozen USD is released as well.
  ; Now allocate the final amounts owed after applying the trade and fees:
  ; Instead of owing Alice 1 full BTC, we now owe her 99,000 USD (she sold her BTC).
  ; This reduces her frozen BTC liability and increases her available USD liability.
  Liabilities:Users:Alice:Wallets:Available:0     -99000.00 USD
  ; This means we now owe Alice 99,000 USD instead of 1 BTC.
  ; Instead of owing Bob 100,000 USD, we owe him 0.99 BTC (he bought BTC):
  Liabilities:Users:Bob:Wallets:Available:0     -0.99000000 BTC
  ; We now owe Bob 0.99 BTC instead of 100,000 USD.
  ; Record the fee from Alice in USD as income:
  ; Alice should have received 100,000 USD, but got only 99,000 USD. 
  ; The difference (1,000 USD) is the exchange’s fee income.
  Income:Exchange:Fees                             -1000.00 USD
  ; Record the fee from Bob in BTC as income:
  ; Bob should have received 1.00000000 BTC but got only 0.99000000 BTC.
  ; The 0.01 BTC difference is fee income for the exchange.
  ; However, we must also separate user-owned assets from exchange-owned assets.
  ; Initially, all BTC and USD held by the exchange were owed to users (Liabilities).
  ; Now that we have taken a fee, part of the assets are no longer owed to users; they are the exchange's own.
  ; We first remove the 0.01 BTC from the user-owned custody:
  Assets:Custody:Crypto                         -0.01000000 BTC
  ; Then we add that 0.01 BTC to the exchange’s own crypto account with a cost basis annotation:
  ; This annotates that the exchange acquired this 0.01 BTC at 100,000 USD/BTC = 1,000 USD cost.
  Assets:Exchange:Crypto                         0.01000000 BTC {100000.00 USD}
  ; Similarly, for the USD fee: 
  ; We remove 1,000 USD from the user-owned fiat custody (since it's no longer owed to users):
  Assets:Custody:Fiat                              -1000.00 USD
  ; And we add those 1,000 USD to the exchange’s own fiat account:
  Assets:Exchange:Fiat                              1000.00 USD
  ; Finally, we recognize the BTC-based fee as income (an additional -1000.00 USD worth of income):
  ; The total fee income is now 2,000 USD worth (1,000 USD from Alice + 1,000 USD from Bob’s BTC).
  Income:Exchange:Fees                             -1000.00 USD
  ; In summary:
  ; - We cleared frozen liabilities and reassigned them to the correct final state.
  ; - We recorded fee income for both the USD and BTC portions.
  ; - We moved the fee amounts from user-owned Assets:Custody to the exchange’s own Assets:Exchange accounts.
  ; - The cost basis annotation on the BTC fee ensures that we know the historical cost of the BTC for tax and gain/loss calculations.

```

### What Has Happened After This Trade?
- The exchange’s Liabilities to Alice now reflect that it owes her 99,000 USD (instead of 1 BTC).
- The exchange’s Liabilities to Bob now reflect that it owes him 0.99 BTC (instead of 100,000 USD).
- The exchange’s Income:Exchange:Fees account has recorded a total of 2,000 USD worth of fee income (1,000 USD in USD form and another 1,000 USD worth in BTC form).
- The exchange’s own Assets have increased by 1,000 USD and 0.01 BTC (with a 1,000 USD cost basis), representing the fee it earned.
- The exchange’s Custody assets have decreased correspondingly, as those amounts are no longer owed to the users.
In short, the trade execution transaction resolved the order, distributed proceeds minus fees, and properly recorded the fee as the exchange’s income and private inventory.
